[FEA] Support Databricks 7.3 LTS Runtime

tgravescs opened the pull request:
fixes 502

This adds a Databricks shim for spark 3.0.1 - runtime 7.3.  

Unfortunately I copied a lot of code from the spark 3.0.0 databricks version. I can't build both at the same time since I have to spin up different clusters with different versions.

I cleaned up some of the build scripts to not need as many parameters and just get the data from maven pom files.

Since this is 3.0.1 some of the files also use the shuffleLike classes, which was an attempt to make AQE work. Unfortunately we still can't support that due to other databricks classes.

The integration tests all pass with AQE disabled and I verified the jenkins files work properly by running them from dev. I did file a followup issue to document the databricks limitations and update anything else I may have missed.


<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

tgravescs commented:
build

jlowe commented on jenkins/databricks/run-tests.py:
```diff
@@ -18,6 +18,7 @@
 import time
 import os
 import subprocess
+import shlex
```
Is this import used?

tgravescs replied:
nope, not needed anymore

jlowe commented on shims/spark301db/pom.xml:
```diff
@@ -0,0 +1,117 @@
+                        <configuration>
+                            <target>
+                                <mkdir dir="${project.build.directory}/extra-resources"/>
+                                <exec executable="bash" output="${project.build.directory}/extra-resources/spark-${spark301db.version}-info.properties">
```
This needs the `failonerror` change akin to what is being done in #1062.

tgravescs replied:
ok we will have to figure out which ones goes in first, I'll watch it

jlowe replied:
You don't have to wait for that PR, it will only help to add it in this PR.

jlowe replied:
Would be good to get the `failonerror` attribute change into this PR even if it occurs before the other PR.

jlowe commented on shims/spark301db/src/main/scala/com/nvidia/spark/rapids/shims/spark301db/GpuHashJoin.scala:
```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2020, NVIDIA CORPORATION.
```

jlowe commented on shims/spark301db/src/main/scala/com/nvidia/spark/rapids/shims/spark301db/GpuShuffledHashJoinExec.scala:
```diff
@@ -0,0 +1,152 @@
+          buildIter, localBuildOutput)) { buildBatch: ColumnarBatch =>
+          withResource(GpuProjectExec.project(buildBatch, gpuBuildKeys)) { keys =>
+            val combined = GpuHashJoin.incRefCount(combine(keys, buildBatch))
+            val filtered = filterBuiltTableIfNeeded(combined)
+            combinedSize =
+                GpuColumnVector.extractColumns(filtered)
+                    .map(_.getBase.getDeviceMemorySize).sum.toInt
+            withResource(filtered) { filtered =>
+              GpuColumnVector.from(filtered)
```
Note this was recently updated in #1066, this should be updated to match that.

tgravescs replied:
thanks for catching I'll upmerge and diff all the files again.

jlowe commented on shims/spark301db/src/main/scala/com/nvidia/spark/rapids/shims/spark301db/Spark301dbShims.scala:
```diff
@@ -0,0 +1,220 @@
+
+package com.nvidia.spark.rapids.shims.spark301db
+
+import java.time.ZoneId
```
There are many unused imports in this file.

tgravescs replied:
I'll update

tgravescs commented:
addressed review comments, code compiles on databricks currently running the tests there, putting it up here to run the regular tests in parallel, so please hold off on merging til I say databricks tests are done

tgravescs commented:
build

tgravescs commented:
build

tgravescs commented:
build

tgravescs commented:
build

tgravescs commented:
tests passed on databricks
